File: benchmarks/tsb/bench_groupby_agg.ts
import { DataFrame } from "tsb";

const rng = (seed: number) => { let s = seed; return () => { s = (s * 1664525 + 1013904223) & 0xffffffff; return (s >>> 0) / 0xffffffff; }; };
const rand = rng(42);
const groups = ["A","B","C","D","E"];
const df = new DataFrame({
  group: Array.from({ length: 100_000 }, () => groups[Math.floor(rand() * 5)]),
  val1: Array.from({ length: 100_000 }, () => (rand() * 2 - 1) * 3),
  val2: Array.from({ length: 100_000 }, () => (rand() * 2 - 1) * 3),
});
for (let i = 0; i < 3; i++) df.groupby("group").agg({ val1: ["mean","std","min","max"], val2: ["sum","count"] });
const N = 30;
const t0 = performance.now();
for (let i = 0; i < N; i++) df.groupby("group").agg({ val1: ["mean","std","min","max"], val2: ["sum","count"] });
const elapsed = performance.now() - t0;
console.log(JSON.stringify({ function: "groupby_agg", mean_ms: elapsed / N, iterations: N, total_ms: elapsed }));

File: benchmarks/pandas/bench_groupby_agg.py
import pandas as pd, json, time, numpy as np
rng = np.random.default_rng(42)
df = pd.DataFrame({
    "group": rng.choice(["A","B","C","D","E"], size=100_000),
    "val1": rng.standard_normal(100_000),
    "val2": rng.standard_normal(100_000),
})
for _ in range(3): df.groupby("group").agg({"val1": ["mean","std","min","max"], "val2": ["sum","count"]})
N = 30
t0 = time.perf_counter()
for _ in range(N): df.groupby("group").agg({"val1": ["mean","std","min","max"], "val2": ["sum","count"]})
elapsed = time.perf_counter() - t0
print(json.dumps({"function": "groupby_agg", "mean_ms": elapsed/N*1000, "iterations": N, "total_ms": elapsed*1000}))
